File: ruby/word-count/word_count.rb
class Phrase
  def initialize(string)
    string2 = string.downcase.gsub(/[^a-zA-Z_0-9 ,']/, '')
    puts "string2: #{string2}"
    @words = string2.split(/[ ,]/).delete_if {|word| word.empty?}.map {|word| word[/^'.*'$/] ? word[1...-1] : word}
    puts "words: #{@words}"
  end

  def word_count
    hash = {}
    @words.uniq.each {|word| hash[word] = @words.count(word)}
    hash
  end
end

module BookKeeping
  VERSION = 1 # Where the version number matches the one in the test.
end

File: elixir/word-count/lib/word_count.ex
defmodule WordCount do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    sentence
    |> String.replace(["!","&","@","$","%","^","&",":", ","], "")
    |> String.downcase
    |> String.split([" ", "_"], trim: true)
    |> Enum.reduce(%{}, &(add_word_count/2))
  end

  defp add_word_count(word, acc) do
    Map.update(acc, word, 1, &(&1+1))
  end

end
